src/core/smack.c

```diff
@@ -0,0 +1,953 @@
+
+    int32_t width;
+    int32_t height;
+    int32_t frames;
+    int32_t us_per_frame;
+    int32_t flags;
+    int32_t trees_size;
+    int32_t audio_size[7];
+    int32_t audio_rate[7];
+
+    long frame_data_offset_in_file;
+    long *frame_offsets;
+    int32_t *frame_sizes;
+    uint8_t *frame_types;
+
+    hufftree16 *mmap_tree;
+    hufftree16 *mclr_tree;
+    hufftree16 *full_tree;
+    hufftree16 *type_tree;
+
+    frame_data_t frame_data;
+    int32_t current_frame;
+};
+
+static const uint8_t BIT_MASKS[] = {
+    0x00,
+    0x01,
+    0x03,
+    0x07,
+    0x0f,
+    0x1f,
+    0x3f,
+    0x7f,
+    0xff,
+};
+
+static const uint8_t PALETTE_MAP[64] = {
+    0x00, 0x04, 0x08, 0x0C, 0x10, 0x14, 0x18, 0x1C,
+    0x20, 0x24, 0x28, 0x2C, 0x30, 0x34, 0x38, 0x3C,
+    0x41, 0x45, 0x49, 0x4D, 0x51, 0x55, 0x59, 0x5D,
+    0x61, 0x65, 0x69, 0x6D, 0x71, 0x75, 0x79, 0x7D,
+    0x82, 0x86, 0x8A, 0x8E, 0x92, 0x96, 0x9A, 0x9E,
+    0xA2, 0xA6, 0xAA, 0xAE, 0xB2, 0xB6, 0xBA, 0xBE,
+    0xC3, 0xC7, 0xCB, 0xCF, 0xD3, 0xD7, 0xDB, 0xDF,
+    0xE3, 0xE7, 0xEB, 0xEF, 0xF3, 0xF7, 0xFB, 0xFF
+};
+
+static const int CHAIN_SIZE[64] = {
+     1,    2,    3,    4,    5,    6,    7,    8,
+     9,   10,   11,   12,   13,   14,   15,   16,
+    17,   18,   19,   20,   21,   22,   23,   24,
+    25,   26,   27,   28,   29,   30,   31,   32,
+    33,   34,   35,   36,   37,   38,   39,   40,
+    41,   42,   43,   44,   45,   46,   47,   48,
+    49,   50,   51,   52,   53,   54,   55,   56,
+    57,   58,   59,  128,  256,  512, 1024, 2048
+};
+
+static inline void *smk_malloc(size_t s)
+{
+    void *buf = malloc(s);
+    if (buf) {
+        memset(buf, 0, s);
+    }
+    return buf;
+}
+
+static int32_t read_i32(uint8_t *data)
+{
+    return (int32_t) (data[0] | (data[1] << 8) | (data[2] << 16) | (data[3] << 24));
+}
+
+// Bitstream functions
+
+static bitstream *bitstream_init(bitstream *bs, const uint8_t *data, int len)
+{
+    bs->data = data;
+    bs->length = len;
+    bs->index = 0;
+    bs->bit_index = 0;
+    return bs;
+}
+
+static inline int read_bit(bitstream *bs)
+{
+    if (bs->index >= bs->length) {
+        return 0;
+    }
+    int result = bs->data[bs->index] & (1 << bs->bit_index);
+    if (++bs->bit_index >= 8) {
+        bs->index++;
+        bs->bit_index = 0;
+    }
+    return result ? 1 : 0;
+}
+
+static inline uint8_t read_byte(bitstream *bs)
+{
+    if (bs->bit_index == 0) {
+        // special case: on exact byte boundary
+        if (bs->index < bs->length) {
+            return bs->data[bs->index++];
+        } else {
+            return 0;
+        }
+    }
+    if (bs->index + 1 >= bs->length) {
+        return 0;
+    }
+    uint8_t value = bs->data[bs->index] >> bs->bit_index;
+    bs->index++;
+    value |= (bs->data[bs->index] & BIT_MASKS[bs->bit_index]) << (8 - bs->bit_index);
+    return value;
+}
+
+// 8-bit huffman tree functions
+
+static huffnode8 *build_tree8_nodes(bitstream *bs, hufftree8 *tree, int value, int bits)
```

Comment:
Unless I'm missing something, `int value` is passed on recursively, but never actually used.

Reply:
It's actually a leftover from when I had verbose logging on practically everything. Good catch.